runtime/src/transaction_batch.rs

```diff
@@ -46,12 +46,47 @@ impl<'a, 'b> TransactionBatch<'a, 'b> {
     pub fn needs_unlock(&self) -> bool {
         self.needs_unlock
     }
+
+    /// Release unnecessary locks before execution.
+    pub fn release_locks_early(&mut self, results: impl Iterator<Item = Result<()>>) {
+        // If we don't need to unlock, then we don't need to release locks early
+        if !self.needs_unlock() {
+            return;
+        }
+
+        let Self {
+            ref mut lock_results,
+            bank,
+            ref sanitized_txs,
+            ..
+        } = self;
+
+        let tx_and_is_locked = results
+            .zip(sanitized_txs.iter())
+            .zip(lock_results.iter_mut())
+            .filter(|(_, lock_result)| lock_result.is_ok())
+            .filter(|((result, _), _)| result.is_err())
```

Comment:
Just used any error here, not only fee paying errors.
If we have already caught a failure, I can't think of a reason we'd want to maintain our hold on the locks.